Add smart pointers to CookieStore and CookieJar

abigailfox opened the pull request:
#### 1963873fe29814ca67ea08fe3afea222fa6ad311
<pre>
Add smart pointers to CookieStore and CookieJar
<a href="https://bugs.webkit.org/show_bug.cgi?id=274285">https://bugs.webkit.org/show_bug.cgi?id=274285</a>
<a href="https://rdar.apple.com/128231784">rdar://128231784</a>

Reviewed by Chris Dumez and Ryosuke Niwa.

* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::MainThreadBridge::ensureOnMainThread):
(WebCore::CookieStore::MainThreadBridge::get):
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
(WebCore::CookieStore::eventListenersDidChange):
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::cookies const):
(WebCore::CookieJar::cookieRequestHeaderFieldProxy):
(WebCore::CookieJar::setCookies):
(WebCore::CookieJar::cookiesEnabled):
(WebCore::CookieJar::cookieRequestHeaderFieldValue const):
(WebCore::CookieJar::getRawCookies const):
(WebCore::CookieJar::setRawCookie):
(WebCore::CookieJar::deleteCookie):

Canonical link: <a href="https://commits.webkit.org/279064@main">https://commits.webkit.org/279064@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0d0a7e7a38299a2938680cc727d70c833c93bb0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52207 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31538 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4627 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55481 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2930 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54511 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2628 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42496 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1882 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54303 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29177 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45060 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23566 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26444 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2334 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1089 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48355 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2480 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57077 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27332 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2534 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49886 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28566 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45176 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49125 "2 api tests failed or timed out") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11444 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29478 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28310 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -272,7 +272,7 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
     };
 
-    m_mainThreadBridge->get(WTFMove(options), WTFMove(completionHandler));
+    Ref { m_mainThreadBridge }->get(WTFMove(options), WTFMove(completionHandler));
```
We should add a protectedMainBridge() function that returns a Ref, since this is done several times.

rniwa replied:
Rather than keep repeating `Ref { m_mainThreadBridge }`, we should add protectedMainThreadBridge function which returns a Ref.

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -92,11 +92,11 @@ String CookieJar::cookies(Document& document, const URL& url) const
 
     std::pair<String, bool> result;
-    if (auto* session = m_storageSessionProvider->storageSession())
+    if (WeakPtr session = Ref { m_storageSessionProvider }->storageSession())
```
We should add a `protectedStorageSessionProvider()` since this is done several times.

rniwa commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -98,7 +98,7 @@ void CookieStore::MainThreadBridge::ensureOnMainThread(Function<void(ScriptExecu
     ASSERT(m_cookieStore);
 
-    RefPtr context = m_cookieStore->scriptExecutionContext();
+    RefPtr context = RefPtr { m_cookieStore.get() }->scriptExecutionContext();
```
I think the preferred convention is to add protectedCookieStore which returns a Ref and call that here.

rniwa commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -123,10 +123,10 @@ void CookieJar::setCookies(Document& document, const URL& url, const String& coo
         frameID = frame->loader().frameID();
 
-    if (auto* session = m_storageSessionProvider->storageSession())
+    if (WeakPtr session = Ref { m_storageSessionProvider }->storageSession())
```
Same comment for adding protectedStorageSessionProvider() instead.

rniwa commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -183,10 +183,10 @@ bool CookieJar::getRawCookies(const Document& document, const URL& url, Vector<C
         frameID = frame->loader().frameID();
 
-    if (auto* session = m_storageSessionProvider->storageSession())
+    if (WeakPtr session = Ref { m_storageSessionProvider }->storageSession())
```
We should probably make `NetworkStorageSession` inherit from `CanMakeCheckedPtr` and use a `CheckedPtr` here instead. `WeakPtr` doesn't prevent UAF here because even if `WeakPtr` was not null at the time of if condition evaluation, it can become null later underneath us. `CheckedPtr` will prevent UAF in that case.

charliewolfe commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -123,10 +123,10 @@ void CookieJar::setCookies(Document& document, const URL& url, const String& coo
     std::optional<FrameIdentifier> frameID;
-    if (auto* frame = document.frame())
+    if (RefPtr frame = document.frame())
         frameID = frame->loader().frameID();
```
Couldn't these just be `frameID = frame->frameID();`? Then we wouldn't need the smart pointer.

rniwa commented on Source/WebCore/platform/network/NetworkStorageSession.h:
```diff
@@ -121,8 +121,12 @@ class CookiesEnabledStateObserver : public CanMakeWeakPtr<CookiesEnabledStateObs
+class NetworkStorageSession
+    : public CanMakeWeakPtr<NetworkStorageSession>
+    , public CanMakeCheckedPtr<NetworkStorageSession> {
+        WTF_MAKE_NONCOPYABLE(NetworkStorageSession);
```
Nit: I don't think we normally extra indent macros like this.

abigailfox committed:
Add smart pointers to CookieStore and CookieJar
https://bugs.webkit.org/show_bug.cgi?id=274285
rdar://128231784

Reviewed by Chris Dumez and Ryosuke Niwa.

* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::MainThreadBridge::ensureOnMainThread):
(WebCore::CookieStore::MainThreadBridge::get):
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
(WebCore::CookieStore::eventListenersDidChange):
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::cookies const):
(WebCore::CookieJar::cookieRequestHeaderFieldProxy):
(WebCore::CookieJar::setCookies):
(WebCore::CookieJar::cookiesEnabled):
(WebCore::CookieJar::cookieRequestHeaderFieldValue const):
(WebCore::CookieJar::getRawCookies const):
(WebCore::CookieJar::setRawCookie):
(WebCore::CookieJar::deleteCookie):

Canonical link: https://commits.webkit.org/279064@main
